File: canrisk-bridge/r/R/canrisk_client.R
# Call Evagene's /risk/canrisk endpoint and return the ##CanRisk 2.0 body.

CANRISK_HEADER <- "##CanRisk 2.0"
HTTP_OK_LOWER <- 200L
HTTP_OK_UPPER <- 300L

#' Raise an api_error condition.
#' @noRd
stop_api <- function(message) {
  stop(structure(
    list(message = message, call = sys.call(-1)),
    class = c("api_error", "error", "condition")
  ))
}

#' Raise a canrisk_format_error condition.
#' @noRd
stop_format <- function(message) {
  stop(structure(
    list(message = message, call = sys.call(-1)),
    class = c("canrisk_format_error", "error", "condition")
  ))
}

#' Fetch a pedigree as a ##CanRisk 2.0 file.
#'
#' @param gateway A gateway function built by e.g. `make_httr2_gateway()`.
#' @param base_url Evagene base URL (trailing slash optional).
#' @param api_key Value for the `X-API-Key` header.
#' @param pedigree_id UUID string.
#' @return The response body as a single character string beginning with
#'   `##CanRisk 2.0`.
fetch_canrisk <- function(gateway, base_url, api_key, pedigree_id) {
  url <- sprintf("%s/api/pedigrees/%s/risk/canrisk", sub("/$", "", base_url), pedigree_id)

  response <- gateway(list(
    url = url,
    headers = c("X-API-Key" = api_key, "Accept" = "text/tab-separated-values")
  ))

  if (response$status < HTTP_OK_LOWER || response$status >= HTTP_OK_UPPER) {
    stop_api(sprintf("Evagene API returned HTTP %d for %s", response$status, url))
  }

  if (!startsWith(response$body, CANRISK_HEADER)) {
    stop_format(sprintf(
      paste0(
        "Response body does not begin with '%s'; ",
        "check the pedigree ID and that your key has the 'analyze' scope."
      ),
      CANRISK_HEADER
    ))
  }
  response$body
}

File: canrisk-bridge/node/src/canRiskClient.ts
import type { HttpGateway } from './httpGateway.js';

export const CANRISK_HEADER = '##CanRisk 2.0';

const HTTP_OK_LOWER = 200;
const HTTP_OK_UPPER = 300;

export class ApiError extends Error {
  constructor(message: string) {
    super(message);
    this.name = 'ApiError';
  }
}

export class CanRiskFormatError extends Error {
  constructor(message: string) {
    super(message);
    this.name = 'CanRiskFormatError';
  }
}

export interface CanRiskClientOptions {
  readonly baseUrl: string;
  readonly apiKey: string;
  readonly http: HttpGateway;
}

export class CanRiskClient {
  private readonly baseUrl: string;
  private readonly apiKey: string;
  private readonly http: HttpGateway;

  constructor(options: CanRiskClientOptions) {
    this.baseUrl = options.baseUrl.replace(/\/$/, '');
    this.apiKey = options.apiKey;
    this.http = options.http;
  }

  async fetchForPedigree(pedigreeId: string): Promise<string> {
    const url = `${this.baseUrl}/api/pedigrees/${pedigreeId}/risk/canrisk`;
    const response = await this.http.getText(url, { headers: this.headers() });
    if (response.status < HTTP_OK_LOWER || response.status >= HTTP_OK_UPPER) {
      throw new ApiError(`Evagene API returned HTTP ${response.status.toString()} for ${url}`);
    }

    const body = await response.text();
    if (!body.startsWith(CANRISK_HEADER)) {
      throw new CanRiskFormatError(
        `Response body does not begin with '${CANRISK_HEADER}'; ` +
          `check the pedigree ID and that your key has the 'analyze' scope.`,
      );
    }
    return body;
  }

  private headers(): Record<string, string> {
    return {
      'X-API-Key': this.apiKey,
      Accept: 'text/tab-separated-values',
    };
  }
}
